File: fetch-messages.js

const fs = require('fs');
const path = require('path');

// 读取配置文件
const configPath = '/Users/suchuanlei/.openclaw/openclaw.json';
const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
const appId = config.channels.feishu.appId;
const appSecret = config.channels.feishu.appSecret;

// 飞书 API 基础 URL
const BASE_URL = 'https://open.feishu.cn';

// 获取 tenant_access_token
async function getTenantAccessToken() {
  const res = await fetch(`${BASE_URL}/open-apis/auth/v3/tenant_access_token/internal`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ app_id: appId, app_secret: appSecret })
  });
  const data = await res.json();
  if (data.code !== 0) {
    throw new Error(`获取 tenant_access_token 失败: ${JSON.stringify(data)}`);
  }
  return data.tenant_access_token;
}

// 获取群消息
async function getMessages(tenantAccessToken, chatId, startTime, endTime) {
  const url = new URL(`${BASE_URL}/open-apis/im/v1/messages`);
  url.searchParams.set('container_id_type', 'chat');
  url.searchParams.set('container_id', chatId);
  url.searchParams.set('start_time', Math.floor(startTime / 1000));
  url.searchParams.set('end_time', Math.floor(endTime / 1000));
  url.searchParams.set('page_size', '50');
  url.searchParams.set('sort_type', 'ByCreateTimeAsc');

  const res = await fetch(url.toString(), {
    method: 'GET',
    headers: { 'Authorization': `Bearer ${tenantAccessToken}` }
  });
  const data = await res.json();
  if (data.code !== 0) {
    throw new Error(`获取消息失败: ${JSON.stringify(data)}`);
  }
  return data.data;
}

// 获取用户信息
async function getUserInfo(tenantAccessToken, userId) {
  const res = await fetch(`${BASE_URL}/open-apis/contact/v3/users/${userId}`, {
    method: 'GET',
    headers: { 'Authorization': `Bearer ${tenantAccessToken}` }
  });
  const data = await res.json();
  if (data.code !== 0) {
    return null;
  }
  return data.data.user;
}

// 主函数
async function main() {
  const now = Date.now();
  const oneHourAgo = now - 3600 * 1000;

  const tenantAccessToken = await getTenantAccessToken();
  const messages = await getMessages(
    tenantAccessToken,
    'oc_4e3daa6139b1eb7beebec0de6dcef569',
    oneHourAgo,
    now
  );

  // 获取发送者信息
  const userMap = new Map();
  for (const item of messages.items || []) {
    if (item.sender?.id && !userMap.has(item.sender.id)) {
      const user = await getUserInfo(tenantAccessToken, item.sender.id);
      if (user) {
        userMap.set(item.sender.id, user);
      }
    }
  }

  // 输出结果
  console.log(JSON.stringify({
    messages: messages.items || [],
    users: Object.fromEntries(userMap),
    startTime: oneHourAgo,
    endTime: now
  }, null, 2));
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: workspace/temp/fetch_messages.py
#!/usr/bin/env python3
"""
虾王群消息总结脚本
使用 tenant_access_token 调用飞书 API 获取群消息
"""

import requests
import json
import time
import sys
from datetime import datetime, timedelta

# 配置
APP_ID = "cli_a92027f8bbb85bc6"
APP_SECRET = "qHbubWguafLFd1NppHnXcdIDXf84WXl4"
CHAT_ID = "oc_4e3daa6139b1eb7beebec0de6dcef569"

# 获取 tenant_access_token
def get_tenant_access_token():
    url = "https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal"
    payload = {
        "app_id": APP_ID,
        "app_secret": APP_SECRET
    }
    response = requests.post(url, json=payload)
    data = response.json()
    if data.get("code") != 0:
        raise Exception(f"获取 tenant_access_token 失败: {data}")
    return data["tenant_access_token"]

# 获取群消息
def get_messages(tenant_access_token, chat_id, hours_ago=1):
    url = f"https://open.feishu.cn/open-apis/im/v1/messages"
    headers = {
        "Authorization": f"Bearer {tenant_access_token}"
    }
    params = {
        "container_id_type": "chat",
        "container_id": chat_id,
        "sort_type": "ByCreateTimeDesc",
        "page_size": 50
    }
    response = requests.get(url, headers=headers, params=params)
    data = response.json()
    if data.get("code") != 0:
        raise Exception(f"获取消息失败: {data}")
    return data.get("data", {}).get("items", [])

# 过滤指定时间范围内的消息
def filter_messages_by_time(messages, hours_ago=1):
    cutoff_time = int((datetime.now() - timedelta(hours=hours_ago)).timestamp() * 1000)
    filtered = []
    for msg in messages:
        create_time = int(msg.get("create_time", 0))
        if create_time >= cutoff_time:
            filtered.append(msg)
    return filtered

# 解析消息内容
def parse_message_content(msg):
    msg_type = msg.get("msg_type", "")
    try:
        body = json.loads(msg.get("body", "{}"))
        if msg_type == "text":
            return body.get("content", "")
        return f"[{msg_type}]"
    except:
        return msg.get("body", "")

# 主函数
def main():
    try:
        # 获取 token
        token = get_tenant_access_token()
        
        # 获取消息
        messages = get_messages(token, CHAT_ID)
        
        # 过滤最近 1 小时的消息
        recent_messages = filter_messages_by_time(messages, 1)
        
        if not recent_messages:
            print(json.dumps({"count": 0, "messages": []}, ensure_ascii=False))
            return 0
        
        # 准备输出数据
        result = {
            "count": len(recent_messages),
            "messages": []
        }
        
        for msg in recent_messages:
            content = parse_message_content(msg)
            sender = msg.get("sender", {}).get("sender_id", {}).get("open_id", "未知用户")
            result["messages"].append({
                "sender": sender,
                "content": content,
                "create_time": msg.get("create_time"),
                "message_id": msg.get("message_id")
            })
        
        print(json.dumps(result, ensure_ascii=False, indent=2))
        return 0
        
    except Exception as e:
        print(f"错误: {e}", file=sys.stderr)
        return 1

if __name__ == "__main__":
    sys.exit(main())
